### ISSUE TITLE
[IPEX][Fusion] Linear+Activation 결합 불가 원인

### ISSUE THREAD

**Issue by SeBin7** [SeBin7/onednn-xpu-fusedops-bench#7]
### CPU ISA: AVX2 + DL Boost(BF16 미지원)
- **사실**: BF16 핵심 최적화(프리패킹/특정 커널)는 **AVX512/AMX** 계열에서 효과적.
- **결과**: AVX2에서는 BF16 경로가 **제한적/비활성** → BF16 시도 시 **에러** 또는 **우회 필요**.

### Linear가 oneDNN로 매핑되지 않음
- **현상**: 최소 예제(`nn.Linear → ReLU`)에서도 `inner_product` 로그가 **단 한 줄도 없음**.
- **의미**: 현재 **PyTorch/IPEX/oneDNN/ISA 조합**에서 Linear가 **ATen/MKL 경로**로만 실행.
- **결론**: Activation(GELU/RELU)만 oneDNN primitive로 **따로** 찍히고, **Linear+Activation 결합(post-ops) 부재**.

### XPU
- **FP32**: 동작은 하나, **GELU 결합 흔적 없음**(대개 GEMM+post-ops 수준).
- **BF16**: `cross_entropy/softmax` 인근에서 **런타임 에러(환경 의존)** 발생.  
  → **우회(로스만 FP32)** 가능하지만 **안정성 저하**.

### 툴체인 상성/로그에서 보인 신호
- BN folding 실패, JIT 사용 자제 경고, IPEX 그래프 패턴 미매칭
- “**일부 최적화만 적용**”되고 **결정적인 fusion은 불발**
- `torch.compile(backend="ipex")`, JIT trace/freeze 시도에도 **Linear 결합 로그 미등장**

### oneDNN verbose 패턴
- **기대(붙었을 때)**:  
  `inner_product ... attr:post-ops:eltwise:alg:eltwise_relu/gelu ...` (한 줄)
- **실제**:  
  `eltwise_gelu` 같은 **단독 eltwise만 반복** → **결합 실패 일관**

### 코드 구조 영향
- `ops.py` **런타임 분기 & 커스텀 경로(myops)** 존재  
  → FX/그래프 기반 최적화가 분기에 걸려 **패턴 캡처/재배치 어려움**
- 분기를 꺼도(치환/후킹) **Linear 자체가 oneDNN로 내려가지 않음** → 근본 해결 불가

---

### 이 머신/셋업에서는 **IPEX를 비활성** 권장
- **CPU**: IPEX/Graph **OFF**, **ATen(+MKL)** 경로로 **FP32 안정 운용**이 실용적
- **XPU**: **FP32 고정**, 그래프 탐색 **OFF**(안정/예측 가능). **BF16은 환경 따라 불안정**

---

## 성능을 더 원하면?
1. **CUDA로 전환**: `torch.compile(backend="inductor") + AMP`  
   - Linear+Act 결합이 **성숙/안정**.
2. **커스텀 C++/oneDNN(핫스팟만)**:  
   - FFN(`Linear → Act → Linear(+residual)`)을 **직접 한 연산으로 묶어 호출**  
   - 단, **유지보수 비용 큼**(Autograd 커스텀, FWD/BWD 검증 필요)
3. **툴체인/하드웨어 교체**:  
   - **IPEX 호환 표준 페어**(Python/PyTorch/IPEX 버전 매칭) + **AVX512/AMX CPU**에서 재시도

---

### CPU (안정 프리셋)
**환경 변수**
```bash
export USE_IPEX_OPT=0
export PREFER_IPEX_GRAPH=0
export OMP_NUM_THREADS=8
```

**실행**
```bash
./scripts/run_mlp.sh cpu --precision fp32 --opt sgd ...
```

**이유**  
Linear가 어차피 **ATen** → IPEX의 복잡도 대비 **실속 없음**.

---

### XPU (안정 프리셋)
**환경 변수**
```bash
export USE_IPEX_OPT=1
export PREFER_IPEX_GRAPH=0
export SYCL_DEVICE_FILTER=level_zero:gpu
export ZE_AFFINITY_MASK=0
```

**실행**
```bash
./scripts/run_mlp.sh xpu --precision fp32 --opt sgd ...
```

**이유**  
FP32에서는 잘 돌아가고, **그래프 탐색 이득 < 변동성**.

---

## “붙였는지” 빠른 판정 체크리스트

**성공 기준(CPU/XPU 공통)** — 아래 **중 한 줄**이라도 로그에 있어야 **진짜 결합**:
```text
inner_product ... post-ops:eltwise:alg:eltwise_relu
inner_product ... post-ops:eltwise:alg:eltwise_gelu
convolution  ... post-ops:eltwise:alg:eltwise_relu
```

**실패 패턴**:
```text
eltwise_gelu  # 만 따로 반복, inner_product 자체 부재 → Linear=ATen
```

---

## 만약 “결합”을 꼭 써야 한다면
- **CUDA 머신**에서 **Inductor/Triton** 활용(가장 쉬움)
- **AVX512/AMX CPU + IPEX 호환 버전 페어**에서 재시도(가능성↑)
- **커스텀 연산**: oneDNN의 `inner_product + post-ops(sum/eltwise)`로 **FFN 직결**,  
  Autograd 커스텀 래퍼로 감싸 **Forward/Backward 검증 필수**

---

## Do / Don’t
**Do**
-  **FP32 위주**로 학습/측정, **로그 파일**로 보관(이미 세팅 OK)
-  Optimizer `foreach=True`, `zero_grad(set_to_none=True)`로 **CPU 병렬/메모리 효율** 확보

**Don’t**
-  이 환경에서 **BF16 강제**(예: IPEX weights prepack) — **오류·편차↑**
-  **FX/JIT**로 무리하게 패턴 고정 — 분기/커스텀 경로 때문에 **유지보수 비용만↑**

**Comment by SeBin7** [comment]
## 1) AVX512/AMX vs AVX2 (CPU ISA 차이)

### AVX2
- **개요**: 256-bit 벡터 명령어(ymm 레지스터)로 부동소수점/정수 연산 가속.
- **BF16 지원**: **미지원** → BF16은 소프트 에뮬/혼합 경로라 이점이 작거나 제한적.
- **최적화 영향**: oneDNN/IPEX에서 공격적인 fusion(예: Linear+GELU post-ops), BF16 프리패킹 같은 최적화가 **제한적**.
- **실전 느낌**: “안전/호환성은 좋지만, 최신 저정밀 최적화 이득은 작다.”

### AVX512
- **개요**: 512-bit 벡터 명령어(zmm 레지스터). 서브셋(AVX512F, BW, VL, VNNI 등)에 따라 성능/기능 차이.
- **VNNI**: INT8 dot-product 가속.
- **BF16**: `AVX512_BF16` 서브셋으로 **BF16 FMA 하드웨어 지원** → BF16 학습/추론에서 큰 이득.
- **최적화 영향**: oneDNN의 `brgemm/inner_product + post-ops`가 **더 자주/더 크게** 이득.

### AMX (Advanced Matrix Extensions)
- **개요**: 최신 Xeon의 **타일 기반 매트릭스 연산 가속**.
- **AMX-BF16 / AMX-INT8**: 대형 GEMM/Convolution의 BF16/INT8 성능을 크게 향상.
- **최적화 영향**: oneDNN이 AMX를 활용하면 **brgemm + post-ops 융합 스루풋**이 크게 증가.

### 실무 관점 요약
- **AVX2 머신**: BF16/대형 fusion 이득이 작음 → **FP32 위주**, IPEX/graph 이득 **제한적**.
- **AVX512/AMX 머신**: BF16/INT8/융합 경로가 **활짝 열림** → IPEX+oneDNN **최적화 체감 큼**.

---

## 2) 용어 설명 (짧고 정확하게)

### BN folding (BatchNorm folding)
- **정의**: 학습된 Conv/Linear와 BatchNorm 파라미터(γ, β, μ, σ)를 **수학적으로 합쳐** 하나의 Conv/Linear로 만드는 최적화.
- **효과**: 레이어 수 감소(메모리/런타임 오버헤드↓), **fusion(Conv+Act)** 성공률↑.
- **비고**: 흔히 **eval/inference**에서 적용.

### JIT (TorchScript JIT)
- **정의**: PyTorch 모델을 트레이싱/스크립팅해 **고정 IR**로 변환 후 최적화/런타임 적용.
- **현황**: 과거 IPEX와 궁합이 좋았으나, 최근엔 **`torch.compile(Inductor)`** 경로 권장 추세.
- **장단점**: 그래프 고정으로 백엔드 최적화는 쉬우나, **동적 분기/파이썬 로직**이 많으면 제약.

### DL Boost (Intel Deep Learning Boost)
- **정의**: Intel CPU의 딥러닝 가속 기술 **마케팅 명칭**. 보통 **AVX512 VNNI(INT8 dot-product)** 등을 가리킴.
- **주의**: “DL Boost 있음”이 곧 **BF16 지원**을 의미하진 않음(VNNI와 BF16은 **별개 서브셋**).

### ATen / MKL
- **ATen**: PyTorch의 **기본 텐서/연산 백엔드**(고수준 연산 추상화).
- **MKL**: Intel **Math Kernel Library**. PyTorch의 일부 연산(특히 BLAS/GEMM)이 MKL을 통해 실행 가능.
- **의미**: 로그/행동에서 “Linear가 ATen/MKL로 간다”는 건, **oneDNN/IPEX 커널이 아닌 기본 경로**(종종 MKL GEMM)를 쓴다는 뜻.

### (oneDNN) verbose / `onednn_verbose`
- **정의**: oneDNN 내부에서 어떤 **primitive**(예: `inner_product`, `convolution`, `eltwise`)가 어떤 ISA/커널로 실행됐는지 **한 줄 로그**로 출력.
- **결합(fusion) 확인법**:
  - **융합 성공**:  
    `inner_product ... attr:post-ops:eltwise:alg:eltwise_gelu/relu ...` 처럼 **같은 줄**에 `post-ops`가 붙어 나오면 `Linear+Act`가 **융합**된 것.
  - **융합 실패**:  
    `eltwise_gelu`만 따로 나오고 `inner_product`가 안 보이면 → **비융합**(Linear=ATen/MKL).

---

## 참고: 빠른 판단 가이드
- **AVX2 + BF16**: 하드웨어 미지원 → 소프트/혼합 경로, **효과 제한**.
- **AVX512_BF16 / AMX**: BF16/INT8 본연의 이점을 **직접 체감** 가능.
- **oneDNN 로그**를 항상 켜고(`ONEDNN_VERBOSE=1`) **post-ops 동반 여부**로 **융합 판정**.
